File: 00055_DP__Jump_Game/sol3.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int idx = nums.size() - 1;
        for (int i = nums.size() - 2; i >= 0; i--) {
            if (i + nums[i] >= idx) {
                idx = i;
            }
        }
        return idx == 0;
    }
};

File: 00032_Stack__Longest_Valid_Parentheses/sol3.py
class Solution:

    def longestValidParentheses(self, s: str) -> int:

        if len(s) < 2:
            return 0

        stack = [-1]
        ans = 0
        for idx, ch in enumerate(s):
            if ch == '(':
                stack.append(idx)
                continue

            stack.pop()
            if not stack:    # pop -1
                stack.append(idx)
            else:
                tmp = idx - stack[-1]
                ans = max(ans, tmp)


        return  ans   
        
